### DISCUSSION TITLE
Airdrop: how can gasplayer get the signature(v, r, s) reasonably?

### DISCUSSION THREAD

**Discussion by EthanWalker10** [Cyfrin/foundry-full-course-cu#3047] [Q&A]
Hey guys,
    I have a question: The premise of Gasplayer helping the user claim and pay the gas fee is that it can obtain the user’s signature, but how can this interaction be implemented?
    If Gasplayer is a smart contract, having the user send the signature to Gasplayer seems expensive and meaningless. If Gasplayer is an official agent and uses a traditional backend service to receive the signature information, then calls the claim function using Gasplayer’s address, this avoids the gas fee. However, doesn’t this lead to centralization? Everything is controlled by Gasplayer, and others can't see or verify it. 
     I believe there should be a better way to implement this. I would greatly appreciate any help.

**Comment by EngrPips** [comment]
Hello @EthanWalker10, why does the Gasplayer have to help the user claim? Is this a requirement of your protocol?

**Reply by EthanWalker10** [reply]
Based on what I’ve learned from the course about Airdrops, the purpose of an Airdrop is to help users obtain minted tokens without consuming gas. This was also demonstrated in the course tests. Look at the following code:
```solidity
    function testUsersCanClaim() public {
        uint256 startingBalance = token.balanceOf(user);

        // get the signature
        vm.startPrank(user);
        (uint8 v, bytes32 r, bytes32 s) = signMessage(userPrivKey, user);
        vm.stopPrank();

        // gasPayer claims the airdrop for the user
        vm.prank(gasPayer);
        airdrop.claim(user, amountToCollect, proof, v, r, s);
        uint256 endingBalance = token.balanceOf(user);
        console.log("Ending balance: %d", endingBalance);
        assertEq(endingBalance - startingBalance, amountToCollect);
    }
```

**Reply by EngrPips** [reply]
I don't think the course content pointed out that there will be a single contract to which every user will send their signature to claim a reward, as you have stated above. Yes, you can actually give anyone your signature to help you claim and pay the gas for the transaction, but there was no idea of everybody giving a single contract their signature for claiming.

**Reply by EthanWalker10** [reply]
So, what you mean is that the user sends the signature information to the gasplayer using the traditional communication method (maybe the gasplayer exposes an interface, or maybe the user directly messages the gasplayer), right? I’m just worried that this seems to undermine decentralization, as the gasplayer could selectively decide whether to help the user claim.
On the other hand, in the example test function I provided, the signatures generated by both the gasplayer and the user are within the same function, so they can naturally be used directly in a straightforward manner. However, in a real-world application, I’m a bit unsure about the best way to handle this step.

**Reply by EngrPips** [reply]
gasPayer is not a contract but another user that has the funds and is willing to pay for your transaction to claim.
